src/content/en/fundamentals/media/manipulating/live-effects.md

```diff
@@ -0,0 +1,374 @@
+project_path: /web/_project.yaml
+book_path: /web/fundamentals/_book.yaml
+description: Many of today's most popular apps let you apply filters and effects to images or video. This article talks about how to implement these features on the open web.
+
+{# wf_updated_on: 2017-08-08 #}
+{# wf_published_on: 2017-08-08 #}
+{# wf_blink_components: Blink>WebGL, Blink>Canvas #}
+
+# Live effects for images and video {: .page-title }
+
+{% include "web/_shared/contributors/mscales.html" %}
+
+Many of today's most popular apps let you apply filters and effects to images or video. This article
+talks about how to implement these features on the open web.
+
+The process is basically the same for videos and images, but I'll cover some important
+video considerations at the end. Throughout the article you can assume that 'image' means 'image a or
+single frame of a video'
+
+## How to get at the pixel data for an image
+
+There are 3 basic categories of image manipulation that are common:
+- Pixel effects like contrast, brightness, warmth, sepia tone, saturation.
+- Multi-pixel effects, called convolution filters, like sharpening, edge detection, blur.
+- Whole image distortion, like cropping, skewing, stretching, lens effects, ripples.
+
+All of these involve getting at the actual pixel data of the source image and then creating a new
+image from it, and the only interface for doing that is a canvas.
+
+The really important choice, then, is whether to do the processing on the CPU, with a 2D canvas, or
+on the GPU, with WebGL.
+
+Let's take a quick look at the differences between the two approaches.
+
+### 2D canvas
+This is definitely, by a long way, the simplest of the two options. First you draw the image on the
+canvas.
+
+    const source = document.getElementById('source-image');
+
+    // Create the canvas and get a context
+    const canvas = document.createElement('canvas');
+    const context = canvas.getContext('2d');
+
+    // Set the canvas to be the same size as the original image
+    canvas.width = source.naturalWidth;
+    canvas.height = source.naturalHeight;
+
+    // Draw the image onto the top-left corner of the canvas
+    context.drawImage(theOriginalImage, 0, 0);
+
+Then you get an array of pixel values for the whole canvas.
+
+    const imageData = context.getImageData(0, 0, canvas.width, canvas.height);
+    const pixels = imageData.data;
+
+At this point the `pixels` variable is a `Uint8ClampedArray` with a length of `width * height * 4`.
+Every array element is one byte and every four elements in the array represents the color of one
+pixel. Each of the four elements represents amount of red, green, blue, and alpha (transparency)
+in that order. The pixels are ordered starting from the top-left corner and working left to right
+and top to bottom.
+
+    pixels[0] = red value for pixel 0
```

Comment:
This is being syntax highlighted weirdly. @petele Can we prevent syntax highlighting on blocks somehow?